gto-support-androidx-viewpager2/src/test/java/org/ccci/gto/android/common/androidx/viewpager2/adapter/PrimaryItemChangeObserverTest.kt

```diff
@@ -0,0 +1,198 @@
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.ext.junit.rules.ActivityScenarioRule
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.viewpager2.widget.ViewPager2
+import com.nhaarman.mockitokotlin2.argThat
+import com.nhaarman.mockitokotlin2.doReturn
+import com.nhaarman.mockitokotlin2.isNull
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.spy
+import com.nhaarman.mockitokotlin2.times
+import com.nhaarman.mockitokotlin2.verify
+import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
+import com.nhaarman.mockitokotlin2.verifyZeroInteractions
+import org.junit.Assert.assertNotNull
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.internal.stubbing.answers.ClonesArguments
+import org.mockito.verification.VerificationMode
+import org.robolectric.Shadows.shadowOf
+import org.robolectric.annotation.Config
+import org.robolectric.shadows.ShadowLooper
+
+@RunWith(AndroidJUnit4::class)
+@Config(sdk = [19, 28])
+class PrimaryItemChangeObserverTest {
+    @get:Rule
+    val activityScenario = ActivityScenarioRule(Activity::class.java)
+
+    private lateinit var viewPager: ViewPager2
+    private lateinit var adapter: TestAdapter
+    private lateinit var looper: ShadowLooper
+
+    @Before
+    fun setup() {
+        activityScenario.scenario.onActivity {
+            viewPager = ViewPager2(it).apply { layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, MATCH_PARENT) }
+            it.setContentView(viewPager)
+        }
+        adapter = spy(TestAdapter())
+        looper = shadowOf(Looper.getMainLooper())
+    }
+
+    // region register/unregister behavior
+    @Test
+    fun registerObserver() {
+        viewPager.adapter = adapter
+        assertNotNull(adapter.observer)
+    }
+
+    @Test
+    fun `registerObserver - Hooks attached`() {
+        val viewPager = mock<ViewPager2>()
+        val recyclerView = mock<RecyclerView> { on { parent } doReturn viewPager }
+
+        adapter.onAttachedToRecyclerView(recyclerView)
+        val observer = adapter.observer!!
+        verify(recyclerView).addOnChildAttachStateChangeListener(observer.childAttachStateChangeListener)
+        verify(viewPager).registerOnPageChangeCallback(observer.pageChangeObserver)
+        verifyNoMoreInteractions(viewPager)
+
+        adapter.onDetachedFromRecyclerView(recyclerView)
+        verify(recyclerView).addOnChildAttachStateChangeListener(observer.childAttachStateChangeListener)
+        verify(viewPager).unregisterOnPageChangeCallback(observer.pageChangeObserver)
+        verifyNoMoreInteractions(viewPager)
+    }
+
+    @Test(expected = IllegalArgumentException::class)
+    fun `registerObserver - Fails when adapter doesn't have stable ids`() {
+        adapter.setHasStableIds(false)
+        viewPager.adapter = adapter
+    }
```

Comment:
This test doesn't have any assertions like ```verify...```, ```assertNotNull```, etc, but I assume the ```expected = IllegalArgumentException:class``` counts as one?

Reply:
yeah, the test only passes when it throws an `IllegalArgumentException`, which `viewPager.adapter = adapter` should throw because the adapter doesn't have stable ids